File: ts/src/session/background-session-events.ts
import type {
  RuntimeSessionEvent,
  RuntimeSessionEventLog,
  RuntimeSessionEventType,
} from "./runtime-events.js";

export type NormalizedSessionEventName =
  | "session_created"
  | "session_queued"
  | "executor_starting"
  | "executor_ready"
  | "prompt_queued"
  | "prompt_started"
  | "runtime_event"
  | "artifact_created"
  | "child_session_created"
  | "session_status"
  | "session_completed";

export type NormalizedSessionEventStatus =
  | "queued"
  | "running"
  | "completed"
  | "failed"
  | "canceled"
  | "skipped"
  | "unknown";

export type NormalizedSessionEventSummaryValue = string | number | boolean;

const SAFE_SANDBOX_CAPABILITY_REASONS = new Set([
  "adapter_error",
  "missing_snapshot_ref",
  "unsupported_prebuild_repo_image",
  "unsupported_resolve_tunnel_ports",
  "unsupported_restore",
  "unsupported_snapshot",
  "unsupported_warm",
]);

export interface NormalizedSessionEvent {
  readonly event_id: string;
  readonly session_id: string;
  readonly sequence: number;
  readonly ts: string;
  readonly event: NormalizedSessionEventName;
  readonly source_event_type:
    | RuntimeSessionEventType
    | "artifact"
    | "lifecycle_hook"
    | "sandbox_capability"
    | "session_status";
  readonly status: NormalizedSessionEventStatus;
  readonly title: string;
  readonly payload_summary: Record<string, NormalizedSessionEventSummaryValue>;
}

export interface LifecycleSessionEventInput {
  readonly sessionId: string;
  readonly sequence: number;
  readonly timestamp: string;
  readonly hook: "setup" | "start" | string;
  readonly phase: "started" | "completed" | "failed" | "timeout" | string;
}

export interface SandboxCapabilitySessionEventInput {
  readonly sessionId: string;
  readonly sequence: number;
  readonly timestamp: string;
  readonly capability: string;
  readonly phase: string;
  readonly bootMode: string;
  readonly provider?: string;
  readonly unsupportedPolicy?: string;
  readonly reason?: string;
  readonly degraded?: boolean;
  readonly error?: string;
}

export interface ArtifactCreatedSessionEventInput {
  readonly sessionId: string;
  readonly sequence: number;
  readonly timestamp: string;
  readonly artifactId: string;
  readonly kind: string;
  readonly label?: string;
  readonly url?: string;
  readonly path?: string;
}

export interface SessionStatusEventInput {
  readonly sessionId: string;
  readonly sequence: number;
  readonly timestamp: string;
  readonly status: NormalizedSessionEventStatus;
  readonly reason?: string;
}

export function normalizeBackgroundSessionTimeline(
  log: RuntimeSessionEventLog,
  opts: { childLogs?: readonly RuntimeSessionEventLog[] } = {},
): NormalizedSessionEvent[] {
  const events = [...log.events]
    .sort((left, right) => left.sequence - right.sequence)
    .map(normalizeRuntimeSessionEvent);
  for (const childLog of opts.childLogs ?? []) {
    events.push(
      ...[...childLog.events]
        .sort((left, right) => left.sequence - right.sequence)
        .map((event) => withChildLineage(normalizeRuntimeSessionEvent(event), childLog)),
    );
  }
  return events.sort((left, right) =>
    `${left.ts}\u0000${left.session_id}\u0000${left.sequence}`.localeCompare(
      `${right.ts}\u0000${right.session_id}\u0000${right.sequence}`,
    ),
  );
}

export function normalizeRuntimeSessionEvent(event: RuntimeSessionEvent): NormalizedSessionEvent {
  switch (event.eventType) {
    case "prompt_submitted":
      return baseEvent(event, {
        normalizedEvent: "prompt_started",
        status: "running",
        title: "Prompt started",
        payloadSummary: pickPayload(event.payload, {
          request_id: "requestId",
          role: "role",
        }),
      });
    case "assistant_message":
      return baseEvent(event, {
        normalizedEvent: "runtime_event",
        status: runtimeActionStatus(event.payload, "completed"),
        title: "Assistant message",
        payloadSummary: pickPayload(event.payload, {
          request_id: "requestId",
          role: "role",
        }),
      });
    case "shell_command":
      return baseEvent(event, {
        normalizedEvent: "runtime_event",
        status: runtimeActionStatus(event.payload, "running"),
        title: "Shell command",
        payloadSummary: pickPayload(event.payload, {
          command: "command",
          cwd: "cwd",
          exit_code: "exitCode",
        }),
      });
    case "tool_call":
      return baseEvent(event, {
        normalizedEvent: "runtime_event",
        status: runtimeActionStatus(event.payload, "completed"),
        title: "Tool call",
        payloadSummary: pickPayload(event.payload, {
          tool: "tool",
          name: "name",
        }),
      });
    case "child_task_started":
      return baseEvent(event, {
        normalizedEvent: "child_session_created",
        status: "running",
        title: "Child session created",
        payloadSummary: pickPayload(event.payload, {
          child_session_id: "childSessionId",
          role: "role",
          task_id: "taskId",
        }),
      });
    case "child_task_completed": {
      const canceled = isCanceledPayload(event.payload);
      const failed = readBoolean(event.payload.isError);
      return baseEvent(event, {
        normalizedEvent: "session_status",
        status: canceled ? "canceled" : failed ? "failed" : "completed",
        title: canceled
          ? "Child session canceled"
          : failed
            ? "Child session failed"
            : "Child session completed",
        payloadSummary: pickPayload(event.payload, {
          task_id: "taskId",
        }),
      });
    }
    case "compaction":
      return baseEvent(event, {
        normalizedEvent: "runtime_event",
        status: "completed",
        title: "Compaction recorded",
        payloadSummary: pickPayload(event.payload, {
          summary_artifact_id: "summaryArtifactId",
        }),
      });
  }
}

export function buildLifecycleSessionEvent(
  input: LifecycleSessionEventInput,
): NormalizedSessionEvent {
  const failed = input.phase === "failed" || input.phase === "timeout";
  const completed = input.phase === "completed";
  const skipped = input.phase === "skipped";
  const event =
    completed && input.hook === "start"
      ? "executor_ready"
      : failed || skipped
        ? "session_status"
        : "executor_starting";
  return {
    event_id: `lifecycle:${input.sessionId}:${input.hook}:${input.phase}:${input.sequence}`,
    session_id: input.sessionId,
    sequence: input.sequence,
    ts: input.timestamp,
    event,
    source_event_type: "lifecycle_hook",
    status: skipped ? "skipped" : failed ? "failed" : completed ? "completed" : "running",
    title: `Lifecycle hook ${input.hook} ${input.phase}`,
    payload_summary: { hook: input.hook, phase: input.phase },
  };
}

export function buildSandboxCapabilitySessionEvent(
  input: SandboxCapabilitySessionEventInput,
): NormalizedSessionEvent {
  void input.error;
  const failed = input.phase === "failed" || input.phase === "timeout";
  const completed = input.phase === "completed";
  const skipped =
    input.phase === "unsupported" || input.phase === "skipped" || input.phase === "degraded";
  const event: NormalizedSessionEventName =
    input.phase === "started"
      ? "executor_starting"
      : completed
        ? "executor_ready"
        : "session_status";
  const status: NormalizedSessionEventStatus = failed
    ? "failed"
    : completed
      ? "completed"
      : skipped
        ? "skipped"
        : "running";
  return {
    event_id: `sandbox:${input.sessionId}:${input.capability}:${input.phase}:${input.sequence}`,
    session_id: input.sessionId,
    sequence: input.sequence,
    ts: input.timestamp,
    event,
    source_event_type: "sandbox_capability",
    status,
    title: `Sandbox ${input.capability.replaceAll("_", " ")} ${input.phase}`,
    payload_summary: sanitizeSummary({
      capability: input.capability,
      phase: input.phase,
      boot_mode: input.bootMode,
      provider: input.provider,
      unsupported_policy: input.unsupportedPolicy,
      reason: safeSandboxCapabilityReason(input.reason),
      degraded: input.degraded ?? false,
    }),
  };
}

export function buildArtifactCreatedSessionEvent(
  input: ArtifactCreatedSessionEventInput,
): NormalizedSessionEvent {
  return {
    event_id: `artifact:${input.sessionId}:${input.artifactId}:${input.sequence}`,
    session_id: input.sessionId,
    sequence: input.sequence,
    ts: input.timestamp,
    event: "artifact_created",
    source_event_type: "artifact",
    status: "completed",
    title: "Artifact created",
    payload_summary: sanitizeSummary({
      artifact_id: input.artifactId,
      kind: input.kind,
      label: input.label,
      path: input.path,
      url: input.url,
    }),
  };
}

export function buildSessionStatusEvent(input: SessionStatusEventInput): NormalizedSessionEvent {
  const terminal = new Set<NormalizedSessionEventStatus>([
    "completed",
    "failed",
    "canceled",
    "skipped",
  ]).has(input.status);
  return {
    event_id: `status:${input.sessionId}:${input.status}:${input.sequence}`,
    session_id: input.sessionId,
    sequence: input.sequence,
    ts: input.timestamp,
    event: terminal ? "session_completed" : "session_status",
    source_event_type: "session_status",
    status: input.status,
    title: `Session ${input.status}`,
    payload_summary: sanitizeSummary({ reason: input.reason }),
  };
}

function baseEvent(
  event: RuntimeSessionEvent,
  opts: {
    normalizedEvent: NormalizedSessionEventName;
    status: NormalizedSessionEventStatus;
    title: string;
    payloadSummary: Record<string, NormalizedSessionEventSummaryValue>;
  },
): NormalizedSessionEvent {
  return {
    event_id: event.eventId,
    session_id: event.sessionId,
    sequence: event.sequence,
    ts: event.timestamp,
    event: opts.normalizedEvent,
    source_event_type: event.eventType,
    status: opts.status,
    title: opts.title,
    payload_summary: opts.payloadSummary,
  };
}

function withChildLineage(
  event: NormalizedSessionEvent,
  childLog: RuntimeSessionEventLog,
): NormalizedSessionEvent {
  return {
    ...event,
    payload_summary: {
      ...event.payload_summary,
      ...sanitizeSummary({
        child_session_id: childLog.sessionId,
        parent_session_id: childLog.parentSessionId,
        task_id: childLog.taskId,
        worker_id: childLog.workerId,
      }),
    },
  };
}

function pickPayload(
  payload: Record<string, unknown>,
  mapping: Record<string, string>,
): Record<string, NormalizedSessionEventSummaryValue> {
  const summary: Record<string, NormalizedSessionEventSummaryValue> = {};
  for (const [outputKey, inputKey] of Object.entries(mapping)) {
    const value = payload[inputKey];
    if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
      summary[outputKey] = value;
    }
  }
  return summary;
}

function sanitizeSummary(
  value: Record<string, unknown>,
): Record<string, NormalizedSessionEventSummaryValue> {
  const summary: Record<string, NormalizedSessionEventSummaryValue> = {};
  for (const [key, item] of Object.entries(value)) {
    if (typeof item === "string" || typeof item === "number" || typeof item === "boolean") {
      summary[key] = item;
    }
  }
  return summary;
}

function safeSandboxCapabilityReason(reason: string | undefined): string | undefined {
  return reason && SAFE_SANDBOX_CAPABILITY_REASONS.has(reason) ? reason : undefined;
}

function runtimeActionStatus(
  payload: Record<string, unknown>,
  defaultStatus: NormalizedSessionEventStatus,
): NormalizedSessionEventStatus {
  if (isCanceledPayload(payload)) {
    return "canceled";
  }
  if (hasFailurePayload(payload)) {
    return "failed";
  }
  if (hasCompletedPayload(payload)) {
    return "completed";
  }
  return defaultStatus;
}

function isCanceledPayload(payload: Record<string, unknown>): boolean {
  return [readPhase(payload), readStatus(payload)].some(
    (value) => value === "canceled" || value === "cancelled",
  );
}

function hasFailurePayload(payload: Record<string, unknown>): boolean {
  const exitCode = readNumber(payload.exitCode);
  return (
    readBoolean(payload.isError) ||
    (exitCode !== null && exitCode !== 0) ||
    readNonEmptyString(payload.error) !== "" ||
    new Set(["error", "failed", "failure", "timeout", "timed_out"]).has(readPhase(payload))
  );
}

function hasCompletedPayload(payload: Record<string, unknown>): boolean {
  const exitCode = readNumber(payload.exitCode);
  return (
    exitCode === 0 ||
    new Set(["completed", "complete", "success", "succeeded"]).has(readPhase(payload))
  );
}

function readPhase(payload: Record<string, unknown>): string {
  return readNonEmptyString(payload.phase).toLowerCase().replace(/-/g, "_");
}

function readStatus(payload: Record<string, unknown>): string {
  return readNonEmptyString(payload.status).toLowerCase().replace(/-/g, "_");
}

function readNonEmptyString(value: unknown): string {
  return typeof value === "string" && value.trim() ? value.trim() : "";
}

function readNumber(value: unknown): number | null {
  return typeof value === "number" ? value : null;
}

function readBoolean(value: unknown): boolean {
  return value === true;
}

File: autocontext/src/autocontext/session/background_session_events.py
from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Any, Literal, TypeAlias

from autocontext.session.runtime_events import RuntimeSessionEvent, RuntimeSessionEventLog

NormalizedSessionEventName: TypeAlias = Literal[
    "session_created",
    "session_queued",
    "executor_starting",
    "executor_ready",
    "prompt_queued",
    "prompt_started",
    "runtime_event",
    "artifact_created",
    "child_session_created",
    "session_status",
    "session_completed",
]
NormalizedSessionEventStatus: TypeAlias = Literal["queued", "running", "completed", "failed", "canceled", "skipped", "unknown"]
NormalizedSessionEventSummaryValue: TypeAlias = str | int | float | bool
NormalizedSessionEvent: TypeAlias = dict[str, Any]


def normalize_background_session_timeline(
    log: RuntimeSessionEventLog,
    *,
    child_logs: Sequence[RuntimeSessionEventLog] | None = None,
) -> list[NormalizedSessionEvent]:
    events = [normalize_runtime_session_event(event) for event in sorted(log.events, key=lambda event: event.sequence)]
    for child_log in child_logs or []:
        events.extend(
            _with_child_lineage(normalize_runtime_session_event(event), child_log)
            for event in sorted(child_log.events, key=lambda event: event.sequence)
        )
    return sorted(events, key=lambda event: (str(event["ts"]), str(event["session_id"]), int(event["sequence"])))


def normalize_runtime_session_event(event: RuntimeSessionEvent) -> NormalizedSessionEvent:
    if event.event_type == "prompt_submitted":
        return _base_event(
            event,
            normalized_event="prompt_started",
            status="running",
            title="Prompt started",
            payload_summary=_pick_payload(event.payload, {"request_id": "requestId", "role": "role"}),
        )
    if event.event_type == "assistant_message":
        return _base_event(
            event,
            normalized_event="runtime_event",
            status=_runtime_action_status(event.payload, "completed"),
            title="Assistant message",
            payload_summary=_pick_payload(event.payload, {"request_id": "requestId", "role": "role"}),
        )
    if event.event_type == "shell_command":
        return _base_event(
            event,
            normalized_event="runtime_event",
            status=_runtime_action_status(event.payload, "running"),
            title="Shell command",
            payload_summary=_pick_payload(event.payload, {"command": "command", "cwd": "cwd", "exit_code": "exitCode"}),
        )
    if event.event_type == "tool_call":
        return _base_event(
            event,
            normalized_event="runtime_event",
            status=_runtime_action_status(event.payload, "completed"),
            title="Tool call",
            payload_summary=_pick_payload(event.payload, {"tool": "tool", "name": "name"}),
        )
    if event.event_type == "child_task_started":
        return _base_event(
            event,
            normalized_event="child_session_created",
            status="running",
            title="Child session created",
            payload_summary=_pick_payload(
                event.payload,
                {"child_session_id": "childSessionId", "role": "role", "task_id": "taskId"},
            ),
        )
    if event.event_type == "child_task_completed":
        canceled = _is_canceled_payload(event.payload)
        failed = event.payload.get("isError") is True
        status: NormalizedSessionEventStatus = "canceled" if canceled else "failed" if failed else "completed"
        title = "Child session canceled" if canceled else "Child session failed" if failed else "Child session completed"
        return _base_event(
            event,
            normalized_event="session_status",
            status=status,
            title=title,
            payload_summary=_pick_payload(event.payload, {"task_id": "taskId"}),
        )
    return _base_event(
        event,
        normalized_event="runtime_event",
        status="completed",
        title="Compaction recorded",
        payload_summary=_pick_payload(event.payload, {"summary_artifact_id": "summaryArtifactId"}),
    )


def build_lifecycle_session_event(
    *,
    session_id: str,
    sequence: int,
    timestamp: str,
    hook: str,
    phase: str,
) -> NormalizedSessionEvent:
    failed = phase in {"failed", "timeout"}
    completed = phase == "completed"
    skipped = phase == "skipped"
    normalized_event: NormalizedSessionEventName
    if completed and hook == "start":
        normalized_event = "executor_ready"
    elif failed or skipped:
        normalized_event = "session_status"
    else:
        normalized_event = "executor_starting"
    return {
        "event_id": f"lifecycle:{session_id}:{hook}:{phase}:{sequence}",
        "session_id": session_id,
        "sequence": sequence,
        "ts": timestamp,
        "event": normalized_event,
        "source_event_type": "lifecycle_hook",
        "status": "skipped" if skipped else "failed" if failed else "completed" if completed else "running",
        "title": f"Lifecycle hook {hook} {phase}",
        "payload_summary": {"hook": hook, "phase": phase},
    }


def build_artifact_created_session_event(
    *,
    session_id: str,
    sequence: int,
    timestamp: str,
    artifact_id: str,
    kind: str,
    label: str | None = None,
    url: str | None = None,
    path: str | None = None,
) -> NormalizedSessionEvent:
    return {
        "event_id": f"artifact:{session_id}:{artifact_id}:{sequence}",
        "session_id": session_id,
        "sequence": sequence,
        "ts": timestamp,
        "event": "artifact_created",
        "source_event_type": "artifact",
        "status": "completed",
        "title": "Artifact created",
        "payload_summary": _sanitize_summary(
            {"artifact_id": artifact_id, "kind": kind, "label": label, "path": path, "url": url}
        ),
    }


def build_session_status_event(
    *,
    session_id: str,
    sequence: int,
    timestamp: str,
    status: NormalizedSessionEventStatus,
    reason: str | None = None,
) -> NormalizedSessionEvent:
    terminal = status in {"completed", "failed", "canceled", "skipped"}
    return {
        "event_id": f"status:{session_id}:{status}:{sequence}",
        "session_id": session_id,
        "sequence": sequence,
        "ts": timestamp,
        "event": "session_completed" if terminal else "session_status",
        "source_event_type": "session_status",
        "status": status,
        "title": f"Session {status}",
        "payload_summary": _sanitize_summary({"reason": reason}),
    }


def _base_event(
    event: RuntimeSessionEvent,
    *,
    normalized_event: NormalizedSessionEventName,
    status: NormalizedSessionEventStatus,
    title: str,
    payload_summary: dict[str, NormalizedSessionEventSummaryValue],
) -> NormalizedSessionEvent:
    return {
        "event_id": event.event_id,
        "session_id": event.session_id,
        "sequence": event.sequence,
        "ts": event.timestamp,
        "event": normalized_event,
        "source_event_type": event.event_type.value,
        "status": status,
        "title": title,
        "payload_summary": payload_summary,
    }


def _with_child_lineage(
    event: NormalizedSessionEvent,
    child_log: RuntimeSessionEventLog,
) -> NormalizedSessionEvent:
    payload_summary = dict(event["payload_summary"])
    payload_summary.update(
        _sanitize_summary(
            {
                "child_session_id": child_log.session_id,
                "parent_session_id": child_log.parent_session_id,
                "task_id": child_log.task_id,
                "worker_id": child_log.worker_id,
            }
        )
    )
    return {**event, "payload_summary": payload_summary}


def _pick_payload(
    payload: Mapping[str, Any],
    mapping: Mapping[str, str],
) -> dict[str, NormalizedSessionEventSummaryValue]:
    summary: dict[str, NormalizedSessionEventSummaryValue] = {}
    for output_key, input_key in mapping.items():
        value = payload.get(input_key)
        if isinstance(value, str | int | float | bool):
            summary[output_key] = value
    return summary


def _sanitize_summary(value: Mapping[str, Any]) -> dict[str, NormalizedSessionEventSummaryValue]:
    summary: dict[str, NormalizedSessionEventSummaryValue] = {}
    for key, item in value.items():
        if isinstance(key, str) and isinstance(item, str | int | float | bool):
            summary[key] = item
    return summary


def _runtime_action_status(
    payload: Mapping[str, Any],
    default_status: NormalizedSessionEventStatus,
) -> NormalizedSessionEventStatus:
    if _is_canceled_payload(payload):
        return "canceled"
    if _has_failure_payload(payload):
        return "failed"
    if _has_completed_payload(payload):
        return "completed"
    return default_status


def _is_canceled_payload(payload: Mapping[str, Any]) -> bool:
    return _read_phase(payload) in {"canceled", "cancelled"} or _read_status(payload) in {"canceled", "cancelled"}


def _has_failure_payload(payload: Mapping[str, Any]) -> bool:
    exit_code = _read_exit_code(payload)
    return (
        payload.get("isError") is True
        or (exit_code is not None and exit_code != 0)
        or _read_non_empty_string(payload.get("error")) != ""
        or _read_phase(payload) in {"error", "failed", "failure", "timeout", "timed_out"}
    )


def _has_completed_payload(payload: Mapping[str, Any]) -> bool:
    exit_code = _read_exit_code(payload)
    return exit_code == 0 or _read_phase(payload) in {"completed", "complete", "success", "succeeded"}


def _read_phase(payload: Mapping[str, Any]) -> str:
    return _read_non_empty_string(payload.get("phase")).lower().replace("-", "_")


def _read_status(payload: Mapping[str, Any]) -> str:
    return _read_non_empty_string(payload.get("status")).lower().replace("-", "_")


def _read_non_empty_string(value: Any) -> str:
    return value.strip() if isinstance(value, str) and value.strip() else ""


def _read_exit_code(payload: Mapping[str, Any]) -> int | None:
    value = payload.get("exitCode")
    return value if isinstance(value, int) else None
